### PR TITLE
feat(x/ecocredit): minimal baskets ORM + keeper setup

### PR DESCRIPTION
**Pull request by aaronc**
## Description

Closes: #735 

* moves code from #698 to x/ecocredit/server/ormutil and sets up the server for minimal baskets
* adds a `basket.Keeper` setup with mocks for `EcocreditKeeper` and `BankKeeper` that can be used for isolated unit testing

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by aaronc** [comment]
Updated this with a keeper setup we can use for unit testing and marking R4R again

File: x/ecocredit/server/basket/keeper.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+// BankKeeper abstracts over methods that the main bank keeper
+// needs to expose to the basket keeper.
+//
+// NOTE: run `make mocks` whenever you add methods here
+type BankKeeper interface {
+	MintCoins(ctx sdk.Context, moduleName string, amt sdk.Coins) error
+	BurnCoins(ctx sdk.Context, moduleName string, amt sdk.Coins) error
+	SendCoinsFromAccountToModule(ctx sdk.Context, senderAddr sdk.AccAddress, recipientModule string, amt sdk.Coins) error
+	SendCoinsFromModuleToAccount(ctx sdk.Context, senderModule string, recipientAddr sdk.AccAddress, amt sdk.Coins) error
+}
```
Comment (ryanchristo) [inline x/ecocredit/server/basket/keeper.go:50]: So we need to include `SpendableCoins` here as well?

File: x/ecocredit/server/keeper.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// In this keeper.go files we expose methods the basket.Keeper needs.
+
+func (s serverImpl) GetCreateBasketFee(ctx context.Context) sdk.Coins {
+	//TODO implement me
+	panic("implement me")
+}
```
Comment (ryanchristo) [inline x/ecocredit/server/keeper.go:14]: Should we add this as a separate item in #732 or is the idea to include it with `CreateBasket`?

**Review by ryanchristo — APPROVED** [review]
Looks good to me.

File: x/ecocredit/server/keeper.go (hunk 2 above)
Comment (aaronc) [reply x/ecocredit/server/keeper.go:14]: the idea is to include with `CreateBasket` but just stub out an example

File: x/ecocredit/server/basket/keeper.go (hunk 1 above)
Comment (aaronc) [reply x/ecocredit/server/basket/keeper.go:50]: maybe, but we can add when it's needed. these are just to start stubbing thing

**Comment by aaronc** [comment]
I stubbed out `MsgServer` implementations just so there are placeholders for those.

Tests are breaking on release/v2 because of the `Params` change. I'll address that in a separate PR

**Review by technicallyty — APPROVED** [review]
lgtm

File: x/ecocredit/server/ormutil/db.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+		wrapper := storeWrapper{store}
+		return ormtable.NewBackend(ormtable.BackendOptions{
+			CommitmentStore: wrapper,
+			IndexStore:      wrapper,
+		}), nil
+	}
+	options.GetBackend = getBackend
+	options.GetReadBackend = func(ctx context.Context) (ormtable.ReadBackend, error) {
+		return getBackend(ctx)
+	}
```
Comment (robert-zaremba) [inline x/ecocredit/server/ormutil/db.go:28]:
```suggestion
	options.GetReadBackend = getBackend
```

File: x/ecocredit/server/ormutil/db_test.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+func sdkContextForStoreKey(key *types.KVStoreKey) sdk.Context {
+	db := dbm.NewMemDB()
+	cms := store.NewCommitMultiStore(db)
+	cms.MountStoreWithDB(key, sdk.StoreTypeIAVL, db)
+	err := cms.LoadLatestVersion()
+	if err != nil {
+		panic(err)
+	}
+	return sdk.NewContext(cms, tmproto.Header{}, false, log.NewNopLogger())
+}
```
Comment (robert-zaremba) [inline x/ecocredit/server/ormutil/db_test.go:27]: I have an impression that I'm  seeing this construction in multiple places (eg in server tests, maybe in other modules as well). Let's maybe move to the root?

File: x/ecocredit/server/ormutil/store.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
+package ormutil
+
+import storetypes "github.com/cosmos/cosmos-sdk/types"
+import "github.com/cosmos/cosmos-sdk/orm/types/kv"
+
+type storeWrapper struct {
```
Comment (robert-zaremba) [inline x/ecocredit/server/ormutil/store.go:6]: how about `errorStore`?

File: x/ecocredit/server/ormutil/db_test.go (hunk 4 above)
Comment (aaronc) [reply x/ecocredit/server/ormutil/db_test.go:27]: what do you mean by root?

File: x/ecocredit/server/ormutil/store.go (hunk 5 above)
Comment (aaronc) [reply x/ecocredit/server/ormutil/store.go:6]: not sure i'm following

File: x/ecocredit/server/ormutil/db_test.go (hunk 4 above)
Comment (aaronc) [reply x/ecocredit/server/ormutil/db_test.go:27]: btw this code already lives in `types/` on master but i moved it here to not have to deal with other breaking changes in `types/`

**Pull request merged**